File: skills/code-conventions/assets/examples/OrderServiceTest.java
// Example: unit test following the code-conventions contract.
// Characterization tests go in a separate permanent class: OrderServiceCharacterizationTest.
class OrderServiceTest {

    private final OrderService service = new OrderService(new FixedClock(OrderFixtures.FROZEN_NOW));

    @Test
    void shouldApplyVolumeDiscountWhenQuantityExceedsThreshold() {
        // Given
        Order order = OrderFixtures.orderWithQuantity(OrderService.VOLUME_DISCOUNT_THRESHOLD + 1);

        // When
        Invoice invoice = service.invoice(order);

        // Then
        assertThat(invoice.discountRate()).isEqualTo(new BigDecimal("0.10"));
    }

    @Test
    void shouldBuildFullInvoiceWhenOrderIsStandard() {
        // Given
        Order order = OrderFixtures.standardOrder();

        // When
        Invoice invoice = service.invoice(order);

        // Then — whole-object assert, never field-by-field cascades
        assertThat(invoice)
                .usingRecursiveComparison()
                .isEqualTo(OrderFixtures.expectedStandardInvoice());
    }
}

File: skills/code-conventions/assets/examples/order-service.test.ts
// Example: Vitest test following the code-conventions contract.
// Characterization tests go in a separate permanent file: order-service.characterization.test.ts.
import { describe, expect, it } from 'vitest'
import { invoice, VOLUME_DISCOUNT_THRESHOLD } from './order-service'
import { expectedStandardInvoice, orderWithQuantity, standardOrder } from './order-fixtures'

describe('invoice', () => {
  it('should apply volume discount when quantity exceeds threshold', () => {
    // Given
    const order = orderWithQuantity(VOLUME_DISCOUNT_THRESHOLD + 1)

    // When
    const result = invoice(order)

    // Then
    expect(result.discountRate).toBe(0.1)
  })

  it('should build full invoice when order is standard', () => {
    // Given
    const order = standardOrder()

    // When
    const result = invoice(order)

    // Then — whole-object assert, never field-by-field cascades
    expect(result).toMatchObject(expectedStandardInvoice())
  })
})
